pkg/arvo/tests/sys/zuse/crypto/aes.hoon

```diff
@@ -0,0 +1,413 @@
+  ::
+    :+  0x8e73.b0f7.da0e.6452.c810.f32b.8090.79e5.
+          62f8.ead2.522c.6b7b
+      0x6bc1.bee2.2e40.9f96.e93d.7e11.7393.172a.
+        ae2d.8a57.1e03.ac9c.9eb7.6fac.45af.8e51.
+        30c8.1c46.a35c.e411
+    0x8a1d.e5be.2eb3.1aad.089a.82e6.ee90.8b0e
+  ::
+    :+  0x8e73.b0f7.da0e.6452.c810.f32b.8090.79e5.
+          62f8.ead2.522c.6b7b
+      0x6bc1.bee2.2e40.9f96.e93d.7e11.7393.172a.
+        ae2d.8a57.1e03.ac9c.9eb7.6fac.45af.8e51.
+        30c8.1c46.a35c.e411.e5fb.c119.1a0a.52ef.
+        f69f.2445.df4f.9b17.ad2b.417b.e66c.3710
+    0xa1d5.df0e.ed79.0f79.4d77.5896.59f3.9a11
+  ==
+++  test-aes-macc
+  %+  do-test-vectors-mac  macc
+  :~
+    :+  0x603d.eb10.15ca.71be.2b73.aef0.857d.7781.1f35.2c07.3b61.08d7.2d98.10a3.0914.dff4
+      0x0
+    0x289.62f6.1b7b.f89e.fc6b.551f.4667.d983
+  ::
+    :+  0x603d.eb10.15ca.71be.2b73.aef0.857d.7781.1f35.2c07.3b61.08d7.2d98.10a3.0914.dff4
+      0x6bc1.bee2.2e40.9f96.e93d.7e11.7393.172a
+    0x28a7.023f.452e.8f82.bd4b.f28d.8c37.c35c
+  ::
+    :+  0x603d.eb10.15ca.71be.2b73.aef0.857d.7781.1f35.2c07.3b61.08d7.2d98.10a3.0914.dff4
+      0x6bc1.bee2.2e40.9f96.e93d.7e11.7393.172a.
+        ae2d.8a57.1e03.ac9c.9eb7.6fac.45af.8e51.
+        30c8.1c46.a35c.e411
+    0xaaf3.d8f1.de56.40c2.32f5.b169.b9c9.11e6
+  ::
+    :+  0x603d.eb10.15ca.71be.2b73.aef0.857d.7781.1f35.2c07.3b61.08d7.2d98.10a3.0914.dff4
+      0x6bc1.bee2.2e40.9f96.e93d.7e11.7393.172a.
+        ae2d.8a57.1e03.ac9c.9eb7.6fac.45af.8e51.
+        30c8.1c46.a35c.e411.e5fb.c119.1a0a.52ef.
+        f69f.2445.df4f.9b17.ad2b.417b.e66c.3710
+    0xe199.2190.549f.6ed5.696a.2c05.6c31.5410
+  ==
+::
+::  SIV mode. Test vectors from Appendix A of RFC 5297:
+::  https://tools.ietf.org/html/rfc5297#appendix-A
+::
++$  vector-siv  [key=@ux ad=(list @ux) in=@ux out=(trel @ux @ud @ux)]
+::
+++  do-test-vectors-siv
+  |*  [sivx=_siva ves=(list vector-siv)]
+  ^-  tang
+  |^  %+  weld
+        %+  category  "encrypting"
+        (zing (turn ves enc))
+      %+  category  "decrypting"
+      (zing (turn ves dec))
+  ::
+  ++  enc
+    |=  vector-siv
+    %+  expect-eq
+      !>  out
+      !>  (~(en sivx key ad) in)
+  ::
+  ++  dec
+    |=  vector-siv
+    %+  expect-eq
+      !>  `in
+      !>  (~(de sivx key ad) p.out q.out r.out)
+  --
+::
+++  test-aes-siva
+  %+  do-test-vectors-siv  siva
+  ^-  (list vector-siv)
+  :~
+    :^    0xfffe.fdfc.fbfa.f9f8.f7f6.f5f4.f3f2.f1f0.
+            f0f1.f2f3.f4f5.f6f7.f8f9.fafb.fcfd.feff
+        :~  0x1011.1213.1415.1617.1819.1a1b.1c1d.1e1f.
+              2021.2223.2425.2627
+        ==
+      0x1122.3344.5566.7788.99aa.bbcc.ddee
+    :+
+      0x8563.2d07.c6e8.f37f.950a.cd32.0a2e.cc93
+      14
+      0x40c0.2b96.90c4.dc04.daef.7f6a.fe5c
+    ::
+    :^    0x7f7e.7d7c.7b7a.7978.7776.7574.7372.7170.
+            4041.4243.4445.4647.4849.4a4b.4c4d.4e4f
+        :~  0x11.2233.4455.6677.8899.aabb.ccdd.eeff.
+            dead.dada.dead.dada.ffee.ddcc.bbaa.9988.
+            7766.5544.3322.1100
```

Comment:
this test case is currently failing, because this AD has leading zeros. What's the easiest way to prepend leading zeros to a byte string? I noticed that some functions take the length as a separate argument (for example, manually hashing this AD with "maca `40" produces the correct output), but there's no way to supply the length of the AD in the `siva` API.

Also, if I happen to discover any test inputs that cause crashes, should I treat those as security vulnerabilities and report them through some secure channel, or is it fine to post them here?

Reply:
You could skip it, or factor it into a call to the SIV internals where you can pass the length.

In general, you never need to handle leading zeros in urbit, except when integrating with 3rd-party systems. The siv{a|b|c} gates are currently for urbit cryptography, and don't have a 3rd-party integration use-case.